src/ngAnimate/animateQueue.js

```diff
@@ -0,0 +1,544 @@
+      forEach(children, function(child) {
+        var state = parseInt(child.getAttribute(NG_ANIMATE_ATTR_NAME));
+        var details = activeAnimationsLookup.get(child);
+        switch (state) {
+          case RUNNING_STATE:
+            details.runner.end();
+            /* falls through */
+          case PRE_DIGEST_STATE:
+            if (details) {
+              activeAnimationsLookup.remove(child);
+            }
+            break;
+        }
+      });
+    }
+
+    function clearElementAnimationState(element) {
+      element = element.length ? element[0] : element;
+      element.removeAttribute(NG_ANIMATE_ATTR_NAME);
+      activeAnimationsLookup.remove(element);
+    }
+
+    function isMatchingElement(a,b) {
+      a = a.length ? a[0] : a;
+      b = b.length ? b[0] : b;
+      return a === b;
+    }
+
+    function closeParentClassBasedAnimations(startingElement) {
+      var parentNode = startingElement[0];
+      do {
+        if (!parentNode || parentNode.nodeType !== ELEMENT_NODE) break;
+
+        var details = activeAnimationsLookup.get(parentNode);
+        if (details) {
+          examineParentAnimation(parentNode, details);
+        }
+
+        parentNode = parentNode.parentNode;
+      } while (true);
+
+      // since animations are detected from CSS classes, we need to flush all parent
+      // class-based animations so that the parent classes are all present for child
+      // animations to properly function (otherwise any CSS selectors may not work)
+      function examineParentAnimation(node, animationDetails) {
+        // enter/leave/move always have priority
+        if (animationDetails.structural) return;
+
+        if (animationDetails.state === RUNNING_STATE) {
+          animationDetails.runner.end();
+        }
+        clearElementAnimationState(node);
+      }
+    }
+
+    function areAnimationsAllowed(element, parent, event) {
+      var bodyElementDetected = false;
+      var rootElementDetected = false;
+      var parentAnimationDetected = false;
+      var animateChildren;
+
+      while (parent && parent.length) {
+        var parentNode = parent[0];
+        if (parentNode.nodeType !== ELEMENT_NODE) {
+          // no point in inspecting the #document element
+          break;
+        }
+
+        var details = activeAnimationsLookup.get(parentNode) || {};
+        // either an enter, leave or move animation will commence
+        // therefore we can't allow any animations to take place
+        // but if a parent animation is class-based then that's ok
+        if (!parentAnimationDetected) {
+          parentAnimationDetected = details.structural || disabledElementsLookup.get(parentNode);
+        }
+
+        if (isUndefined(animateChildren) || animateChildren === true) {
+          var value = parent.data(NG_ANIMATE_CHILDREN_DATA);
+          if (isDefined(value)) {
+            animateChildren = value;
+          }
+        }
+
```

Comment:
At this point, if `parentAnimationDetected === true` and `animateChildren === false`, we could immediately return false. There is no need to keep looping.